### PR TITLE
telegram simple extension

### PR DESCRIPTION
**Pull request by innightwolfsleep**
Adding telegram bot with simple setup.
- impersonate, continue generate, regenerate, delete previous, reset history buttons,
- local chat history - no loss memory after text-generator restart. Every char have its own separate history.
- characters list for loading
- chat and notebook modes,
- separated session for all users (by chat_id)

### CODE REVIEW

File: extensions/telegram_bot/script.py [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+            stopping_strings = []
+            eos_token = None
+            if self.bot_mode in ["chat", "chat-restricted"]:
+                # don't know why, but better works without stopping_strings
+                # stopping_strings = [f'\n{self.users[chat_id]["name2"]}:', f'\n{self.users[chat_id]["name1"]}:']
+                eos_token = '\n'
+            # Make prompt
+            prompt = self.users[chat_id]["context"] + "\n" + "\n".join(self.users[chat_id]["history"])
+            # Generate!
+            generator = generate_reply(
```
Comment (dvoidus) [inline extensions/telegram_bot/script.py:358]: This is not working with the current main branch, use something like this instead:
```
        # Generate!
        params = {
            "max_new_tokens": 1024,
            "do_sample": True,
            "temperature": 0.72,
            "top_p": 0.73,
            "top_k": 0,
            "typical_p": 1,
            "repetition_penalty": 1.1,
            "encoder_repetition_penalty": 1,
            "min_length": 0,
            "no_repeat_ngram_size": 0,
            "num_beams": 1,
            "penalty_alpha": 0,
            "length_penalty": 1,
            "early_stopping": False,
            "seed": -1,
            "add_bos_token": True,
            "custom_stopping_strings": []
        }

        generator = generate_reply(
            question=prompt, state=params,
            eos_token=eos_token, stopping_strings=stopping_strings
        )
```

Comment (innightwolfsleep) [reply extensions/telegram_bot/script.py:420]: Fixed.

**Review by GalvareyPoco — COMMENTED** [review]
i think the character loading hould be updated to the new format, i have made a dirty implementation based on the chat script, something that could be also implemented is instruct mode

`def load_char_json_file(self, char_file):
        # Copy default user data
        user = deepcopy(self.default_users_data.copy())
        filep = self.characters_dir_path if self.bot_mode != 'instruct' else self.instruct_dir_path
        greeting_field = 'greeting'
        # Try to read char file. If reading fail - return default user data
        try:
            for extension in ["yml", "yaml", "json"]:
                filepath = Path(f'{filep}/{char_file}')
                if filepath.exists():
                    break
            file_contents = open(filepath, 'r', encoding='utf-8').read()
            data = json.loads(
                file_contents) if extension == "json" else yaml.safe_load(file_contents)
            #  load persona and scenario
            if 'your_name' in data and data['your_name'] != '':
                user["name1"] = data['your_name']
            user["name2"] = data['name'] if 'name' in data else data['char_name']

            for field in ['context', 'greeting', 'example_dialogue', 'char_persona', 'char_greeting', 'world_scenario']:
                if field in data:
                    data[field] = replace_character_names(
                        data[field], user["name1"], user["name2"])

            if 'context' in data:
                user["context"] = f"{data['context'].strip()}\n\n"
            elif "char_persona" in data:
                user["context"] = build_pygmalion_style_context(data)
                user["greeting_field"] = 'char_greeting'

            if 'example_dialogue' in data:
                user["context"] += f"{data['example_dialogue'].strip()}\n"
            if greeting_field in data:
                user["greeting"] = data[greeting_field]
            if 'end_of_turn' in data:
                user["end_of_turn"] = data['end_of_turn']
            #  after <START> add char greeting
        except Exception as e:
            print("load_char_json_file", e)
        return user`

**Comment by oobabooga** [comment]
This is very interesting, but I will not be able to maintain this extension if it gets merged. I encourage you to upload it to its own GitHub repository. If you do that, I can link it in the extensions page in the wiki:

https://github.com/oobabooga/text-generation-webui/wiki/Extensions

**Comment by innightwolfsleep** [comment]
> This is very interesting, but I will not be able to maintain this extension if it gets merged. I encourage you to upload it to its own GitHub repository. If you do that, I can link it in the extensions page in the wiki:
> 
> https://github.com/oobabooga/text-generation-webui/wiki/Extensions


Let it be so. I made repo for this extension:
https://github.com/innightwolfsleep/text-generation-webui-telegram_bot
Check and link it to wiki, please.

**Comment by oobabooga** [comment]
Added here https://github.com/oobabooga/text-generation-webui/wiki/Extensions

**Comment by voxitme** [comment]
> Added here https://github.com/oobabooga/text-generation-webui/wiki/Extensions

Hey @oobabooga!
Is there any reason it isn't listed anymore? I had difficulties finding it here in a closed issue.

Thank you so much for your work by th way!!

**Pull request closed without merging**